[ONNX] Align behavior of ReduceL2-11, 13, 18 with original framework

LucaTamSapienza opened the pull request:
### Details:
- I've aligned the ReduceL2 operation with opset 11, 13, and 18. I have some doubts about how to implement support for the tensor bfloat16 for opset 13 and also some doubts about opset 18. I've registered the function inside the ops_bridge.cpp, created test models, and added them inside onnx_import.in.cpp.
### Tickets:
 - Closes #20560

p-wysocki commented:
Hello @LucaTamSapienza, is this PR ready for review?

LucaTamSapienza commented:
> Hello @LucaTamSapienza, is this PR ready for review?

Hi @p-wysocki. The code compiles, but I have some pending issues. I have described my problems in the PR specification. I was waiting for someone to provide me with some clarifications on the doubts I encountered. I took a look at [softmax.cpp](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/frontend/src/op/softmax.cpp) file and to [this](https://github.com/openvinotoolkit/openvino/issues/20563) conversation to see how to implement the features of the different versions. Unfortunately, I still have some doubts about `opset13`, on how to support the `bfloat` input. And I'm not too sure about my implementation of `opset11`. If you could take a look at my code and give me some feedback, it would be greatly appreciated. Thanks!

gkrivor commented on src/frontends/onnx/frontend/src/op/reduce.cpp:
```diff
@@ -93,24 +93,28 @@ std::shared_ptr<ov::Node> get_reduction_axes_from_attr(const Node& node) {
 
     const auto reduction_axes = axes_as_attr ? get_reduction_axes_from_attr(node) : get_reduction_axes_from_input(node);
     if (reduction_axes != nullptr) {
+        // Check that axes are within the specified range
```
I think you may not need such overcomplication after rebasing to a fresh master.

LucaTamSapienza replied:
> I think you may not need such overcomplication after rebasing to a fresh master.

Ok, thanks for the advice!

gkrivor commented on src/frontends/onnx/frontend/src/op/reduce.cpp:
```diff
@@ -158,6 +162,30 @@ ov::OutputVector reduce_sum_square(const ov::frontend::onnx::Node& node) {
+
+namespace set_11 {
+ov::OutputVector reduce_l2(const Node& node) {
+    if (node.get_ov_inputs().at(0).get_partial_shape().rank().is_static()) {
```
Not sure it is needed. I suggest to add some missing checks for types. Checks should be aligned with original ONNX documentation.

gkrivor commented on src/frontends/onnx/frontend/src/op/reduce.cpp:
```diff
@@ -158,6 +162,30 @@ ov::OutputVector reduce_sum_square(const ov::frontend::onnx::Node& node) {
+
+namespace set_18 {
+ov::OutputVector reduce_l2(const ov::frontend::onnx::Node& node) {
+    return {make_ng_reduction_op<v4::ReduceL2>(node, node.get_ov_inputs().at(0))};
```
Overall - good, just add missing checks

LucaTamSapienza replied:
> missing checks

I'm checking [ONNX documentation](https://onnx.ai/onnx/operators/onnx__ReduceL2.html#reducel2) to understand what additional checks I should perform. I don't understand which ones you're referring to. Could you please be more specific about it?

gkrivor reviewed (changes requested):
Hi @LucaTamSapienza, thanks for your contribution!

Please, rebase your solution to a fresh master, because I've merged a change https://github.com/openvinotoolkit/openvino/pull/22462 with fix for all Reduce* operations which may affect your test cases.

Also, could you enable tests here https://github.com/openvinotoolkit/openvino/blob/96b72d0c823a44cab1ad89d9e6ca0d0e3ac75c0c/src/frontends/onnx/tests/tests_python/test_backend.py#L485

They should work after rebasing to a fresh master

LucaTamSapienza commented:
> Please, rebase your solution to a fresh master, because I've merged a change #22462 with fix for all Reduce* operations which may affect your test cases.

Sure, i'll sync && pull immediatly.

> Also, could you enable tests here

Do you mean for v 11, 13 ,18?

gkrivor commented:
Hi, I refactored reduce.cpp and reduce.hpp. I think it should be more simple to add a new functionality.

Please, rebase after merge.

https://github.com/openvinotoolkit/openvino/pull/23429

LucaTamSapienza commented:
> Hi, I refactored reduce.cpp and reduce.hpp. I think it should be more simple to add a new functionality.
> 
> Please, rebase after merge.
> 
> #23429

Thanks for it, I see that it has been merged. I'll work on it.

gkrivor commented:
Hi @LucaTamSapienza ,

Please take a look on https://github.com/openvinotoolkit/openvino/pull/23475 as a reference.

Especially take a look on test cases. For example, as I can see you've missed in opset-18 prototxt mention of an "axes" input. It is very important to verify this way.

Also check changes in onnx/tests/__init__.py and onnx/tests/tests_python/test_backend.py

LucaTamSapienza commented:
> Please take a look on #23475 as a reference.

@gkrivor ,

Thank you very much for your help and linking me to this PR, I have properly aligned versions 13 and 18. I noticed that you left a comment under version 11. Perhaps I should remove the call from reduce.cpp file and simply use the same implementation used in v1 by calling it in reduce.hpp within v11? Something like 
```
namespace set_11 {
using set_1::reduce_l2;
} // namespace set_11
```



> Especially take a look on test cases. For example, as I can see you've missed in opset-18 prototxt mention of an "axes" input. It is very important to verify this way.

For this correction, I took inspiration from the prototxt of the `reduceMax` operation in `set_18`. Now should be good.

> Also check changes in onnx/tests/**init**.py and onnx/tests/tests_python/test_backend.py

I've deleted the ReduceL2 references from ticket 99962 && 99968 inside the __init__.py and removed for 99968 declaration about reduce_l2. I just have some doubts about this `"OnnxBackendNodeModelTest.test_reduce_l2_empty_set_cpu"`, should I leave this?

gkrivor commented:
Hi @LucaTamSapienza, thanks for the update!

Regarding opset-11. Actually, I think it is a good idea, I also thought about something like this. But currently, we have an internal discussion about "following the standards". I suggest to keep it as is. When we will decide how strictly we want to support standards - I could easily extend/reduce code base.

Also, I don't recommend to touch this place because several developers are working on reduce* operations right now and you may introduce additional merge-conflicts between you and them.

About  "OnnxBackendNodeModelTest.test_reduce_l2_empty_set_cpu" - depends on a test result. If it become passing (you will see in test results), and leave it if it is still xfailed.

LucaTamSapienza commented:
> Also, I don't recommend to touch this place because several developers are working on reduce* operations right now and you may introduce additional merge-conflicts between you and them.

You're right @gkrivor , I've just resolved some merge conflicts I had, now everything should be fine.

> About "OnnxBackendNodeModelTest.test_reduce_l2_empty_set_cpu" - depends on a test result. If it become passing (you will see in test results), and leave it if it is still xfailed.

I ran `/bin/arm64/Release/ov_onnx_frontend_tests --gtest_filter=\*reduce_l2\*` and all the tests passed successfully.

gkrivor commented on src/frontends/onnx/frontend/src/op/reduce.cpp:
```diff
@@ -186,25 +186,28 @@ ov::OutputVector reduce_sum_square(const ov::frontend::onnx::Node& node) {
-    2. Opset 11 introduced requirement for axes values to be in a range [-r, r-1] where r = rank(data)
-       Same time Reduce* operations in OpenVINO has same requirement from first version
-*/
+namespace set_11 {
```
Please, do not introduce opset-11 right now, everything else looks good

LucaTamSapienza replied:
> Please, do not introduce opset-11 right now, everything else looks good

Sure, I'll remove it also from the header file

gkrivor commented:
build_jenkins

gkrivor commented:
build_jenkins

gkrivor commented:
build_jenkins
